test/utils/test_common.py

```diff
@@ -70,3 +76,90 @@ def test_int(self, expected):
     def test_tuple(self, expected):
         actual = to_tuple_tuple((2, 2, 2), 3, 3)
         assert actual == expected, "tuple -> tuple[tuple] incorrect"
+
+
+class TestCheckpointWrapper:
+    @pytest.fixture
+    def device(self):
+        return torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+
+    @pytest.fixture
+    def model(self, device):
+        class DummyAttention(nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.param = nn.Parameter(torch.ones(3))
+
+            def forward(self, x, y, attn_mask=None, use_cache=None):
+                u = x if use_cache else y
+                u = u * self.param
+                if attn_mask is not None:
+                    u = u * attn_mask
+                return u
+
+        class DummyModel(nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.attention = DummyAttention()
+                self.identity = nn.Identity()
+
+            @checkpoint_wrapper
+            def _layer_to_wrap(self, x, y, attn_mask, use_cache):
+                return self.attention(x, y, attn_mask, use_cache)
+
+            def forward(self, x, y, attn_mask=None, use_cache=False):
+                out = self._layer_to_wrap(x, y, attn_mask, use_cache=use_cache)
+                out = self.identity(out)
+                return out
+
+        return DummyModel().to(device)
+
+    @pytest.fixture
+    def inputs(self, device):
+        # Set inputs to be requires_grad = True or else it will break autograd
+        # during the backward pass,
+        # "RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn"
```

Comment:
Good question here.

- Short answer: 
Checkpointed layers run in `torch.no_grad()` mode but to compute gradients in the backward pass either the input or the output of the checkpointed layer needs to have `grad_fn`.

- Long answer: 
In order for checkpointing to compute the gradient during the backward pass, the input to the first checkpointed layer needs to have `grad_fn` so that the checkpointed layers are not disconnected from autograd. While this condition is almost guaranteed if the checkpointed layers are the middle layers, in this unit test `DummyAttention` is the first layer of the model so I had to make its inputs `requires_grad`. 

Also that checkpointing runs in `torch.no_grad()` model: if the inputs are not attached to gradient, neither will the outputs. In that case, the "RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn" will be thrown.

- Update
Adding a dummy identity layer before the checkpointed attention layer so that inputs to the later will have gradient recorded.

Reply:
Thanks for the explanation. At first glance the use of grads for the input layer is definitely counterintuitive, but now I see why we would need it based on how grads get handled by checkpointing. Your solution to prepend with an Identity layer sounds good to me